### PR TITLE
Port 1.15.2 Villager optimizations

### PR DESCRIPTION
**Pull request by MrIvanPlays**
~~Tested & works without issues.
Also added some more obfuscation helpers for easier maintenance & update between major releases~~
raids don't work with it

### CODE REVIEW

File: Spigot-Server-Patches/0551-Optimize-Villagers.patch
Comment (kennytv) [inline Spigot-Server-Patches/0551-Optimize-Villagers.patch]: _ah yes, the vacany_

Comment (MrIvanPlays) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: Whoops, typos 
Had to reimplement the whole patch by hand :eyes:

**Comment by JanTuck** [comment]
So, you are saying this provides proper behaviour that is consistent with vanilla? Or at least close enuf.

**Comment by MrIvanPlays** [comment]
> So, you are saying this provides proper behaviour that is consistent with vanilla? Or at least close enuf.

It is 100% consistent like aikar described it originally. Couldn't find any inconsistencies with the testing I and other 2 people did. 
You can test it for yourself if you want

File: Spigot-Server-Patches/0551-Optimize-Villagers.patch
Comment (kennytv) [inline Spigot-Server-Patches/0551-Optimize-Villagers.patch]: no

Comment (MrIvanPlays) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: Why waste cpu time on even initializing the loop? 
Sure, as you wish

Comment (BillyGalbreath) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: The compiler optimizes it out

**Comment by Proximyst** [comment]
Does this take into account the new behaviour of "stealing" workstations in 1.16?

**Comment by MrIvanPlays** [comment]
> Does this take into account the new behaviour of "stealing" workstations in 1.16?

Haven't tested that, please do me a favor and test it.

File: Spigot-Server-Patches/0551-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0551-Optimize-Villagers.patch]: 9 spaces?

Comment (MrIvanPlays) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: Um wtf happened here? I remember intellij autocompleting that

Comment (Proximyst) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: Intellisense doesn't apply to whitespace...

Comment (MrIvanPlays) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: bruh
whatever now's been fixed

File: Spigot-Server-Patches/0551-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0551-Optimize-Villagers.patch]: It's cool to want attribution, but you didn't co-author this patch.

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0551-Optimize-Villagers.patch
Comment (MrIvanPlays) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: I want to see this commented from a papermc member otherwise it would stay.

Comment (JRoy) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: Multiple people have ported patches over to new major MC versions countless times. They've never placed their name inside the patch for credit.

Your "port" is no different in that regard, git commit history is here for that.

Comment (mikroskeem) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: Do not slap your name on things you haven't done. This is also the reason why you were banned from the IRC or Paper's Discord Guild.

Comment (duplexsystem) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: I think his confusion comes from @Proximyst saying "If it was based on someone else's but you didn't write the base, it's still their patch. You'd write you co-authored it in that case." In response to Ivan saying "The Pathfinder changes don't exactly match aikar's changes in the previous patch."

Comment (MrIvanPlays) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: pretty sure the reason behind why you don't want it there is because i'm banned from paper discord and you don't want to see me as a contributor. as i said i want to see this commented from papermc member (prefferably, the author of the patch - aikar) and then ill remove it.

> This is also the reason why you were banned from the IRC or Paper's Discord Guild.

It wasn't for that, no one minded I stole 1 patch of your microcord which later I removed. Now I did change and im trying to show it off by starting to contribute things and being more active in the overall minecraft discords, but the paper discord - no one wants to talk with me in other discords and i can't show it off to them.

> Multiple people have ported patches over to new major MC versions countless times. They've never placed their name inside the patch for credit.

valid point, but still want to see it commented from a papermc member.

Comment (pop4959) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: Why are you still arguing about this when it's clear that you're wrong? It doesn't make sense to wait for Aikar to tell you the exact same thing 5 people have already told you. I don't think anyone cares whether you are banned on the Discord or not, contributing is great, and everyone loves you for that. Just don't insist that you put your name on something that isn't your work.

Comment (BillyGalbreath) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: > you don't want to see me as a contributor

Uh.. Even just with 1 commit you will end up in the list of contributors.. https://github.com/PaperMC/Paper/graphs/contributors

Best to handle things the right way, you're just making a clown out of yourself how you're doing it now.

Comment (MrIvanPlays) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: > and everyone loves you for that

nice joke. i still think you're hating me no matter what. even when mini was porting paper to 1.16.1 and started a test server some1 just banned me for no reason. you all fucking hate me, you can stop trying to "show-off" that you actually care about the patch. you only care me to not be a contributor and to NOT participate in anything paper-related (and just for the record, this is NOT towards papermc members, cuz I know some of them do hate me but even then still talk with me, this is towards all of you who started the whole argument).

Comment (jacobsandersen) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: Ivan, can you stop making a fool of yourself for 3 minutes?

You didn't author nor coauthor the patch, Aikar did. You updated it. They are asking you to remove your coauthor attribution because you are not the coauthor of the patch.

This is not in any way different from you choosing any book, updating it, and then slapping your name next to the author. It wouldn't be true and neither is this.

You're free to contribute. If you weren't, you'd be banned from even making issues/PRs so seriously, stop.

You claim all day and night that you've changed, but you literally have not. You love to spout off about how we're targeting you but you are doing specific actions repeatedly that do not align with Paper's goals. If you don't stop, then yeah you're going to get banned and nobody is going to like your attitude.

But that doesn't preclude you from still taking part of the community if you change your ways.

My 2c. Whatever. I doubt anyone cares so much about this patch that they are willing to do drama over it. 

TLDR remove the coauthor status or close the PR, but stop being so annoying.

Comment (MiniDigger) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: Brb, adding myself as co-author to the ~200 patches I updated to 1.16

File: Spigot-Server-Patches/0551-Optimize-Villagers.patch
Comment (aikar) [inline Spigot-Server-Patches/0551-Optimize-Villagers.patch]: This line has to be dropped. It's the only thing different in 1.16 (decreaseVacancy). They no longer claim the workstation upon targeting it, only when reaching it.

File: Spigot-Server-Patches/0551-Optimize-Villagers.patch
Comment (aikar) [inline Spigot-Server-Patches/0551-Optimize-Villagers.patch]: why was this changed?

**Review by aikar — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0551-Optimize-Villagers.patch
Comment (aikar) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: There is nothing significant changed here to warrant a co author. So please remove.

Authorship is more about ideas and implementation design, not the syntax of a line.

Comment (Proximyst) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: It was already removed with https://github.com/PaperMC/Paper/pull/4101/commits/ce978bbff0d937df7a966474779a4ea4c52bbabb

File: Spigot-Server-Patches/0551-Optimize-Villagers.patch
Comment (MrIvanPlays) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: To fit with your changes. See your original filtering (entry.getLongValue() is now entry.getValue().b)

File: Spigot-Server-Patches/0551-Optimize-Villagers.patch
Comment (MrIvanPlays) [reply Spigot-Server-Patches/0551-Optimize-Villagers.patch]: Done

File: Spigot-Server-Patches/0174-revert-serverside-behavior-of-keepalives.patch
Comment (MrIvanPlays) [inline Spigot-Server-Patches/0174-revert-serverside-behavior-of-keepalives.patch]: I have 0 idea of what happened here, pretty sure it's ok?

**Review by aikar — DISMISSED** [review]
@PaperMC/development-team please spot check behavior of villages and claiming workstations before merge, but seems fine.

**Comment by MrIvanPlays** [comment]
I'm not entirely sure which change made this behavior, I have some more in-depth testers of this patch and other patches, and it seems like it causes villagers to not see doors if they're more than 1 block apart from them considering that they face the door exactly, otherwise they completely ignore it.

**Comment by aikar** [comment]
This patch does not relate to doors, this is about POI's and doors are not POI's.

**Comment by MrIvanPlays** [comment]
well then this is an other bug have to go deeper to find what's causing it

**Comment by MrIvanPlays** [comment]
apparently it is causing raids to not trigger ( issue 3941 )

VillagePlace#f(long) uses RegionFileSelection#c(long) which returns empty Optional for some reason, causing VillagePlace#a#b(long) to return 7 which is then returned by WorldServer#b(SectionPosition) used in WorldServer#a(BlockPosition, int) causing it to return false which is used in WorldServer#a_(BlockPosition) which is used in the bad omen tick in order to determine if it should start a raid and I have NO idea how to fix this. 
I'm in the village with bad omen and raid doesn't trigger, fortunately on latest paper which does not have this patch works.
@aikar daddy plz fix

File: Spigot-Server-Patches/0565-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0565-Optimize-Villagers.patch]: Where'd the `decreaseVacancy` call go? We still need to make it clear that it's *not* available, do we not?

Comment (MrIvanPlays) [reply Spigot-Server-Patches/0565-Optimize-Villagers.patch]: image

**Comment by Proximyst** [comment]
Here's the fix to your issue with raids:

```patch
From 70c360533e3a5693e0c688d60eb09b90477d1932 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Wed, 2 Sep 2020 19:10:36 +0200
Subject: [PATCH] Fix to your issue

---
 .../0565-Optimize-Villagers.patch                | 16 +++++++++++++---
 1 file changed, 13 insertions(+), 3 deletions(-)

diff --git a/Spigot-Server-Patches/0565-Optimize-Villagers.patch b/Spigot-Server-Patches/0565-Optimize-Villagers.patch
index cdff4e6c4..6330a91ca 100644
--- a/Spigot-Server-Patches/0565-Optimize-Villagers.patch
+++ b/Spigot-Server-Patches/0565-Optimize-Villagers.patch
@@ -39,9 +39,18 @@ So we cache that and only rebuild it if professions change, which should be neve
 a plugin manipulates and adds custom professions, which it will handle by rebuilding.
 
 diff --git a/src/main/java/net/minecraft/server/BehaviorFindPosition.java b/src/main/java/net/minecraft/server/BehaviorFindPosition.java
-index 63a761ebef80d4af09cdc2682e496d78492c4a3a..466180158919c4bf9fde54791511702b36181f29 100644
+index 63a761ebef80d4af09cdc2682e496d78492c4a3a..bab411b102355bc32ea1d2119604d8d363f6ba66 100644
 --- a/src/main/java/net/minecraft/server/BehaviorFindPosition.java
 +++ b/src/main/java/net/minecraft/server/BehaviorFindPosition.java
+@@ -16,7 +16,7 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
+     private final VillagePlaceType b;
+     private final MemoryModuleType<GlobalPos> c;
+     private final boolean d;
+-    private final Optional<Byte> e;
++    private final Optional<Byte> e; private Optional<Byte> getPoiAcquisitionEvent() { return this.e; } // Paper - OBFHELPER
+     private long f;
+     private final Long2ObjectMap<BehaviorFindPosition.a> g;
+ 
 @@ -48,7 +48,7 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
          if (this.d && entitycreature.isBaby()) {
              return false;
@@ -51,7 +60,7 @@ index 63a761ebef80d4af09cdc2682e496d78492c4a3a..466180158919c4bf9fde54791511702b
              return false;
          } else {
              return worldserver.getTime() >= this.f;
-@@ -56,12 +56,56 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
+@@ -56,12 +56,57 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
      }
  
      protected void a(WorldServer worldserver, EntityCreature entitycreature, long i) {
@@ -91,6 +100,7 @@ index 63a761ebef80d4af09cdc2682e496d78492c4a3a..466180158919c4bf9fde54791511702b
 +                            if (pathentity != null && pathentity.canReach()) {
 +                                GlobalPos globalPos = GlobalPos.create(worldserver.getDimensionKey(), pos);
 +                                entitycreature.getBehaviorController().setMemory(c, globalPos);
++                                this.getPoiAcquisitionEvent().ifPresent((event) -> worldserver.broadcastEntityEffect(entitycreature, event));
 +                                break OUT;
 +                            }
 +                            if (poiAttempts >= maxPoiAttempts) {
@@ -110,7 +120,7 @@ index 63a761ebef80d4af09cdc2682e496d78492c4a3a..466180158919c4bf9fde54791511702b
          Predicate<BlockPosition> predicate = (blockposition) -> {
              BehaviorFindPosition.a behaviorfindposition_a = (BehaviorFindPosition.a) this.g.get(blockposition.asLong());
  
-@@ -103,6 +147,7 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
+@@ -103,6 +148,7 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
              }
          }
  
-- 
2.28.0

```
https://owo.whats-th.is/22eR7E1.patch

Apply: `git apply name-here.patch`.

**Comment by aikar** [comment]
Was this something I missed in what the stream code did in my migration for 1.15?

**Comment by Proximyst** [comment]
I *believe* so, but raids worked fine, so I don't entirely know what happened...

**Comment by MrIvanPlays** [comment]
will commit with rebasing so it is ready for merge. busy rn but in around an hour everything will be ready

**Comment by MrIvanPlays** [comment]
ok github thank you for rejecting a review which was approving the pr...
@aikar do your magic

**Comment by Proximyst** [comment]
You don't have to ping him. The dev team (currently Mini & I) can also approve and merge, and we will once it's been tested thoroughly ^^

**Comment by MrIvanPlays** [comment]
proxi ur fix fixed sth other which was unnoticed, raids are yet still bork

**Comment by MrIvanPlays** [comment]
After countless hours of researching, I'm once again confirming, this patch causes raids to not be working. 
What I've tried to do so far: revert changes in RegionFileSection - no luck
Im going to update on what I tried to do in order to fix the issue.

**Comment by Proximyst** [comment]
I wasn't in a position to answer when you first commented on my patch, and had forgotten it; sorry about that.

Raids *do* work with my patch applied from my testing; what's your specific method of getting a raid to start?

**Comment by MiniDigger** [comment]
and more importantly, is this patch the only one you are running?

**Comment by MrIvanPlays** [comment]
> I wasn't in a position to answer when you first commented on my patch, and had forgotten it; sorry about that.
> 
> Raids _do_ work with my patch applied from my testing; what's your specific method of getting a raid to start?

find a village using /locate, walk to it (or fly), around 100 blocks before that if I was in creative set myself to survival, give myself bad omen, walk inside the village.

> and more importantly, is this patch the only one you are running?

yes. indeed I have 1 more guy that is testing with me too, and he also got the same results.

**Comment by MrIvanPlays** [comment]
Looking at mojang's new 1.16 code, they started sorting positions by distance, so here's my idea:
build something flexible inside their current code to make sure if for some reason the villager's going to push farther POI that the chunk which is going to be tried is loaded, but the fact they sort positions by distance and limiting them by the first 5 I think it's very unlikely for a villager to push POI which's chunks isn't already loaded.

**Comment by aikar** [comment]
All of that is already done. I sorted it by distance in 1.15. The code iterates loaded POI.

Not sure how any of that is suppose to relate to Bad omen.

**Comment by MrIvanPlays** [comment]
> All of that is already done. I sorted it by distance in 1.15. The code iterates loaded POI.
> 
> Not sure how any of that is suppose to relate to Bad omen.

Yeah this wasnt related to raids. 
For raids: I was able to nail down everything to a poi data check, which in my case fails which doesn't start the raid. Removing this patch fixed everything. And its not because poi data was not saved or loaded, that was working fine cuz I was looking at the files with NBTExplorer.

Here's what I found: VillagePlace#a#c(long) returns 7 rather than 0
In the Bad omen tick there's a check, WorldServer#a_ from all redirects to other methods the final thing goes to VillagePlace#a#c(long) . 
I'm not sure why that failed, cuz clearly the poi data was there (at least in the saved files)

back to my last reply: idk what was it in 1.15, but in 1.16 mojang changed it. now it uses VillagePlace#b , where that sort is.

File: Spigot-Server-Patches/0576-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0576-Optimize-Villagers.patch]: `this.b` is what? Obfhelper.

File: Spigot-Server-Patches/0576-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0576-Optimize-Villagers.patch]: `SectionPosition.a` is what? Obfhelper.

File: Spigot-Server-Patches/0576-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0576-Optimize-Villagers.patch]: `this.g` is what? Obfhelper.

File: Spigot-Server-Patches/0576-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0576-Optimize-Villagers.patch]: `this.g` is what? Obfhelper.

File: Spigot-Server-Patches/0576-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0576-Optimize-Villagers.patch]: `this.b` is what? Obfhelper.

File: Spigot-Server-Patches/0576-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0576-Optimize-Villagers.patch]: What is `c`?

File: Spigot-Server-Patches/0576-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0576-Optimize-Villagers.patch]: Please use readable names; `e` tells us nothing, especially in the middle of obfuscated code.

File: Spigot-Server-Patches/0576-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0576-Optimize-Villagers.patch]: `b`? `f`?

File: Spigot-Server-Patches/0576-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0576-Optimize-Villagers.patch]: `this.g` is what? Obfhelper. I know this isn't modified, but it will help when updating the code.

File: Spigot-Server-Patches/0576-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0576-Optimize-Villagers.patch]: `this.g` is what? Obfhelper. I know this isn't modified, but it will help when updating the code.

File: Spigot-Server-Patches/0576-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0576-Optimize-Villagers.patch]: `this.c` is what? Obfhelper.

File: Spigot-Server-Patches/0576-Optimize-Villagers.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0576-Optimize-Villagers.patch]: `this.c` is what? Obfhelper.

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0576-Optimize-Villagers.patch
Comment (MrIvanPlays) [reply Spigot-Server-Patches/0576-Optimize-Villagers.patch]: This is mojang's original code. I doubt you need something changed here.

**Comment by MrIvanPlays** [comment]
looks like git decided to ditch the old file instead of marking changes

**Comment by MrIvanPlays** [comment]
raids still don't wanna work with that patch. They work on already generated villages, but they don't on a newly generated ones.

**Comment by KisaragiEffective** [comment]
bump

**Comment by MiniDigger** [comment]
@MrIvanPlays do you plan on picking up work on the raids issue here again?

**Comment by Proximyst** [comment]
#5322 Closing due to author being blocked.

**Pull request closed without merging**
